Add Function to Preprocess Request Prompt & Address String Escape Symbol Issue

davidgxue opened the pull request:
### Description
- See https://github.com/astronomer/ask-astro/issues/278 for details of why the error occurs. This won't be explained again in this PR.
- After investigation, it seems like LangChain's implementation, specifically how Langchain calls Weaviate's hybrid retriever is not optimal.  Specifically, see this discussion thread from Weaviate https://forum.weaviate.io/t/hybrid-search-resulting-in-graphql-error/539. Technically, the safer way to build the query is using `.build()` instead of `.do()`. Using `build()` would have prevented symbols such as `\` in the original string to be properly transformed into GraphQL syntax compatible string. However, LangChain implemented this part of the code using `.do()` which causes issues with certain symbols that conflict with the GraphQL syntax. See here https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/retrievers/weaviate_hybrid_search.py#L152 on line 152. 
- As a result, we can only work around this issue by getting rid of the problem causing string escape symbol `\`. We can achieve this by parsing any single standalone `\` used to do string escape out of the original query. This shouldn't affect the original meaning of the prompt as the `\` gets auto added downstream anyway when querying the LLMs with string escape needed so that is not a concern.

### Technical Changes
- Add a `_preprocess_request` function that does input validation and transformation on the user prompt to workaround this issue in LangChain.

### Tests
1. When question exceeds character limit
![image](https://github.com/astronomer/ask-astro/assets/26350341/2c4d7999-72af-4afe-b152-82937d487063)

2. When thread has more 6 questions/answers pairs, checked langsmith and saw only most recent 5 pairs 

3. Questions with more than 1 `\` is kept the same and questions with a single `\` the `\` is removed.
For example, the original github issue's error causing query is now being successfully executed.
Enter prompt: `KeyError: \'gcp\'` (note in Langsmith logs the prompt it auto adds another backslack to existing backslashs`)

![image](https://github.com/astronomer/ask-astro/assets/26350341/05a43540-9a83-4a91-aad3-75aa6a8e7454)


closes #278

Lee-W commented on api/ask_astro/services/questions.py:
```diff
@@ -27,6 +28,20 @@ async def _update_firestore_request(request: AskAstroRequest) -> None:
+        error_msg = "Question text is too long. Please try making a new thread and shortening your question."
+        request.response = error_msg
+        raise Exception(error_msg)
+    if len(request.prompt) == 0:
```
```suggestion
    if not request.prompt:
```

Lee-W commented on api/ask_astro/services/questions.py:
```diff
@@ -27,6 +28,20 @@ async def _update_firestore_request(request: AskAstroRequest) -> None:
+    if len(request.prompt) > 20000:
+        error_msg = "Question text is too long. Please try making a new thread and shortening your question."
+        request.response = error_msg
+        raise Exception(error_msg)
```
I might suggestion adding a custom exception instead of using `Execption`

sunank200 replied:
+1

sunank200 commented on api/ask_astro/services/questions.py:
```diff
@@ -27,6 +28,20 @@ async def _update_firestore_request(request: AskAstroRequest) -> None:
+        request.response = error_msg
+        raise Exception(error_msg)
+    if len(request.messages) > 10:
+        request.messages = request.messages[-10:]
```
why -10?

sunank200 replied:
Should this be a parameter instead so that it can be changed as part of env in deployment?

davidgxue replied:
Fixed!

davidgxue commented:
Addressed your comment about custom error type in the new commit. I changed up the error propagation logic a little bit as well to make it cleaner. Thanks for the help reviewing @Lee-W  @sunank200

Lee-W commented on api/ask_astro/services/questions.py:
```diff
@@ -27,6 +28,20 @@ async def _update_firestore_request(request: AskAstroRequest) -> None:
+    if not request.prompt:
+        error_msg = "Question text cannot be empty. Please try again with a different question."
+        request.response = error_msg
+        raise Exception(error_msg)
```
Could we use a custom exception instead of using Execption here?

davidgxue replied:
Oops my bad I forgot to push my changes haha. I was like I thought I fixed this. Please feel free to re-review and resolve this comment if it's good!
